lupa/_lupa.pyx

```diff
@@ -1327,26 +1383,43 @@ cdef int py_function_result_to_lua(LuaRuntime runtime, lua_State *L, object o) e
      if runtime._unpack_returned_tuples and isinstance(o, tuple):
          push_lua_arguments(runtime, L, <tuple>o)
          return len(<tuple>o)
+     check_lua_stack(L, 1)
      return py_to_lua(runtime, L, o)
 
 cdef int py_to_lua_handle_overflow(LuaRuntime runtime, lua_State *L, object o) except -1:
-    cdef int nargs
-
-    lua.lua_pushlstring(L, LUPAOFH, len(LUPAOFH))
-    lua.lua_rawget(L, lua.LUA_REGISTRYINDEX)
-    if lua.lua_isnil(L, -1):
-        lua.lua_pop(L, 1)
-        return 0
-    nargs = py_to_lua_custom(runtime, L, o, 0)
-    if nargs <= 0:
-        lua.lua_pop(L, 1)
-        return 0
-    if lua.lua_pcall(L, nargs, 1, 0):
-        lua.lua_pop(L, 1)
-        return 0
-    return 1
+    """Converts Python object to Lua via overflow handler
+    Preconditions:
+        Lua runtime is locked
+    Postconditions:
+        Returns 0 if cannot convert Python object to Lua (handler not registered or failed)
+        Returns 1 if the Python object was converted successfully and pushed onto the stack
+    """
+    check_lua_stack(L, 2)
+    old_top = lua.lua_gettop(L)
+    try:
+        lua.lua_pushlstring(L, LUPAOFH, len(LUPAOFH))
+        lua.lua_rawget(L, lua.LUA_REGISTRYINDEX)
+        if lua.lua_isnil(L, -1):
+            lua.lua_pop(L, 1)
+            return 0
+        py_to_lua_custom(runtime, L, o, 0)
+        if lua.lua_pcall(L, 1, 1, 0):
```

Comment:
Are you sure that this can never put more than one item on the stack? I think that used to be possible.

Reply:
Although the current implementation of `py_to_lua_custom` always pushes 1 item, I agree that we should make it generic.

Reply:
Actually, the postcondition for `py_to_lua_custom` is that it pushes one value to the Lua stack and that it returns 1 (except errors).